vesting/lib.rs

```diff
@@ -0,0 +1,267 @@
+#![cfg_attr(not(feature = "std"), no_std, no_main)]
+
+
+#[ink::contract]
+mod vesting_contract {
+
+    #[ink(storage)]
+    pub struct VestingContract {
+        releasable_balance: Balance,
+        released_balance: Balance,
+        duration_time: Timestamp,
+        start_time: Timestamp,
+        beneficiary: AccountId,
+        owner: AccountId,
+    }
+
+    // Error for when the beneficiary is a zero address.
+    // & Error for when the releasable balance is zero.
+    #[derive(Debug, PartialEq, Eq, scale::Encode, scale::Decode)]
+    #[cfg_attr(feature = "std", derive(scale_info::TypeInfo))]
+    pub enum Error {
+        InvalidBeneficiary,
+        ZeroReleasableBalance,
+    }
+
+    // To emitted events when a release is made.
+    #[ink(event)]
+    pub struct Released {
+        value: Balance,
+        to: AccountId,
+    }
+
+    /// # This is to set the following during contract deployment:
+    ///     - beneficiary: the account that will receive the tokens
+    ///     - duration_time: duration of the vesting period,
+    ///       please note that this is in seconds
+    ///     - start_time: the time (as Unix time) at which point
+    ///       vesting starts
+    ///     - owner: the account that can release the tokens
+    ///     - releasable_balance: the initial amount of tokens vested
+    ///     - released_balance: the initial amount of tokens released
+    ///
+    /// # Note:
+    /// The beneficiary cannot be the zero address.
+    ///
+    impl VestingContract {
+        #[ink(constructor, payable)]
+        pub fn new(beneficiary: AccountId, duration_time_in_sec: Timestamp) -> Result<Self, Error> {
+            if beneficiary == AccountId::from([0x0; 32]) {
+                return Err(Error::InvalidBeneficiary);
+            }
+            let duration_time = duration_time_in_sec * 1000;
+            let start_time = Self::env().block_timestamp();
```

Comment:
I feel like this is better be implemented in terms of block numbers. Given average block time of ~6s, let's say, we can calculate the duration of vesting period in the number of block.

Reply:
Hi @SkymanOne  
Thanks for the feedback, I definitely see your point and it makes sense. 
However, I've once [came across this article](https://consensys.github.io/smart-contract-best-practices/development-recommendations/solidity-specific/timestamp-dependence/#avoid-using-blocknumber-as-a-timestamp) that recommends timestamps for time-dependent logic, as I did with `vesting_schedule`; Could you please take a look at the article and let me know if you still prefer that I use block numbers instead?

Let me know